### PR TITLE
Better error message for nonarray indexing

### LINKED ISSUE #2197: Better error message
```fortran
program test
    type :: int_container
        integer :: my_int
    end type
    integer :: my_int

    type(int_container) :: t
    t%my_int(:) = 1
end program
```
This gives:
```console
$ lfortran c.f90
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 2201
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 901
    result = fe.get_asr2(input, lm, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/fortran_evaluator.cpp", line 251
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/fortran_evaluator.cpp", line 273
    compiler_options.symtab_only, compiler_options);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 108
    if (res.ok) {
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 3047
    b.is_body_visitor = true;
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 216
    visit_ast(*x.m_items[i]);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4768
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4725
    case astType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4771
    void visit_mod(const mod_t &b) { visit_mod_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4368
    case modType::BlockData: { v.visit_BlockData((const BlockData_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 1411
    transform_stmts(body, x.n_body, x.m_body);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 187
    this->visit_stmt(*m_body[i]);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4813
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4469
    case stmtType::Assign: { v.visit_Assign((const Assign_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 1987
    this->visit_expr(*x.m_target);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4864
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4531
    case exprType::Compare: { v.visit_Compare((const Compare_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 5035
    tmp = create_ArrayRef(x.base.base.loc, x.m_args, x.n_args, v_expr, v, f2);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 2777
    m_end = ASRUtils::get_bound(v_Var, i + 1, "ubound", al);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/asr_utils.h", line 3674
    " is invalid. Rank of the array, " + std::to_string(arr_n_dims));
LCompilersException: Dimension 1 is invalid. Rank of the array, 0
```
The error message should say that `t%my_int` is not an array, so cannot be indexed.

GFortran's error message is not great either:
```console
$ gfortran c.f90
c.f90:8:5:

    8 |     t%my_int(:) = 1
      |     1
Error: Unclassifiable statement at (1)
```
But at least it's an error message.

### PR DESCRIPTION
**Pull request by implicitall**
Closes #2197 

This adds an error message for indexing a non-array variable or type member.

### CODE REVIEW

**Comment by implicitall** [comment]
Also, this adds an error message for missing lower bound on an array.

**Comment by certik** [comment]
Awesome, thanks for that @implicitall. Let's also add a test case for this into `tests/errors/`.

**Comment by implicitall** [comment]
I was also trying to add an error message for a similar case but found that it might have more side effects than I intended for this pull request so I backed out the change. This could be opened as a separate issue.
`
program test
integer :: a(:)
a(:) = 1
end program
`

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -2775,7 +2775,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                                     nullptr));
                     }
                 } else {
-                    m_end = ASRUtils::get_bound(v_Var, i + 1, "ubound", al);
+                    m_end = ASRUtils::get_bound<SemanticError>(v_Var, i + 1, "ubound", al);
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:2778]: Ahh nice, I was unable to figure out how to do this. Thank you @implicitall!

**Review by Pranavchiku — APPROVED** [review]
Looks good to me! Thanks for fixing this!

**Review by czgdp1807 — APPROVED** [review]
Thanks for this.

**Comment by certik** [comment]
@czgdp1807 you should squash or rebase PRs that have merges from master.

**Comment by czgdp1807** [comment]
@certik I did squash and merge. Please check https://github.com/lfortran/lfortran/commit/fe579db50b07cb34712c4bd284fa57e224509c76 for the same.

**Comment by certik** [comment]
@czgdp1807 awesome, thanks! For some reason I thought I saw two parents, but there is indeed just one. Sorry about that.

**Pull request merged**